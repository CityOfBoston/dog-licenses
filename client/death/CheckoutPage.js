// @flow

import React from 'react';
import Head from 'next/head';
import { observer } from 'mobx-react';

import type Cart from '../store/Cart';
import Nav from '../common/Nav';
import CheckoutItem from './checkout/CheckoutItem';

export type InitialProps = {|
|}

export type Props = {
  /* :: ...InitialProps, */
  cart: Cart,
}

@observer
export default class CheckoutPage extends React.Component {
  props: Props;

  handleSubmit = (ev: SyntheticInputEvent) => {
    ev.preventDefault();
  }

  render() {
    const { cart } = this.props;

    return (
      <div>
        <Head>
          <title>Boston.gov — Death Certificate Checkout</title>
        </Head>

        <Nav cart={cart} link="lookup" />

        <div className="p-a300 b--g">
          <div className="sh sh--b0 m-v300">
            <h1 className="sh-title">Checkout</h1>
          </div>
        </div>

        <form className="" acceptCharset="UTF-8" method="get" action="/death" onSubmit={this.handleSubmit}>
          { cart.items.map((item) => <CheckoutItem key={item.id} item={item} cart={cart} />) }

          <div className="p-a300 b--g">
            <button className="btn" type="submit">Pay and Finish</button>
          </div>
        </form>
      </div>
    );
  }
}
