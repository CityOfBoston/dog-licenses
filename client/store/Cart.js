// @flow

import { observable, computed, action } from 'mobx';

import type { DeathCertificate } from '../types';

export class CartItem {
  @observable.ref cert: DeathCertificate;
  @observable quantity: number;
}

export default class Cart {
  @observable items = [];

  @computed get size(): number {
    return this.items.reduce((acc, item) => acc + item.quantity, 0);
  }

  @action add(cert: DeathCertificate, quantity: number) {
    const existingItem = this.items.find((item) => item.cert.id === cert.id);
    if (existingItem) {
      existingItem.quantity += quantity;
    } else {
      const item = new CartItem();
      item.cert = cert;
      item.quantity = quantity;

      this.items.push(item);
    }
  }

  @action setQuantity(certId: string, quantity: number) {
    const existingItem = this.items.find((item) => item.cert.id === certId);
    if (existingItem) {
      existingItem.quantity = quantity;
    }
  }

  @action remove(certId: string) {
    const idx = this.items.findIndex(({ cert }) => cert.id === certId);
    if (idx !== -1) {
      this.items.splice(idx, 1);
    }
  }
}
