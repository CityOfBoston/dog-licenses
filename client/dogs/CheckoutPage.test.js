// @flow
import React from 'react';
import { runInAction } from 'mobx';
import renderer from 'react-test-renderer';

import Cart from '../store/Cart';
import {
  TYPICAL_LICENSE,
  PENDING_LICENSE,
} from '../../fixtures/client/dog-licenses';

import CheckoutPage from './CheckoutPage';

describe('rendering', () => {
  let cart;

  beforeEach(() => {
    cart = new Cart();
  });

  it('renders loading for unhydrated cart', () => {
    runInAction(() => {
      cart.pendingFetches = 2;
      cart.items = ([
        {
          id: TYPICAL_LICENSE.id,
          cert: null,
          quantity: 5,
        },
        {
          id: PENDING_LICENSE.id,
          cert: null,
          quantity: 3,
        },
      ]: any);
    });

    expect(
      renderer.create(<CheckoutPage cart={cart} />).toJSON(),
    ).toMatchSnapshot();
  });

  it('renders a hydrated cart', () => {
    cart.add(TYPICAL_LICENSE, 5);
    cart.add(PENDING_LICENSE, 2);

    expect(
      renderer.create(<CheckoutPage cart={cart} />).toJSON(),
    ).toMatchSnapshot();
  });
});
