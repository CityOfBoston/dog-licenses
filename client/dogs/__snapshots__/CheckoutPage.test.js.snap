// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders a hydrated cart 1`] = `
<div>
  <nav
    className="nv-s nv-s--sticky"
    data-jsx={792433154}
  >
    <div
      className="nv-s-l bar"
      data-jsx={792433154}
    >
      <a
        className="nv-s-l-b back-link"
        data-jsx={792433154}
        href="/dogs"
        onClick={[Function]}
      >
        Back to Lookup
      </a>
      <a
        className="cart-link"
        data-jsx={792433154}
        href="/dogs/checkout"
        onClick={[Function]}
      >
        7
      </a>
    </div>
  </nav>
  <div
    className="p-a300"
  >
    <div
      className="sh sh--b0 m-v300"
    >
      <h1
        className="sh-title"
      >
        Checkout
      </h1>
    </div>
  </div>
  <div>
    <div
      className="p-a300 result"
      data-jsx={3916862138}
    >
      <input
        aria-label="Quantity"
        className="br br-a150 quantity-box"
        data-jsx={3916862138}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        value="5"
      />
      <div
        className="license-info"
        data-jsx={3916862138}
      >
        <div
          data-jsx={3916862138}
        >
          Bruce
           
          Banner
        </div>
        <div
          data-jsx={3916862138}
          style={
            Object {
              "fontStyle": "italic",
            }
          }
        >
          3/4/2016
          <span
            className="id-label"
            data-jsx={3916862138}
          >
            ID:
          </span>
           
          000002
        </div>
      </div>
      <button
        data-jsx={3916862138}
        onClick={[Function]}
        type="button"
      >
        X
      </button>
    </div>
    <div
      className="p-a300 result"
      data-jsx={3916862138}
    >
      <input
        aria-label="Quantity"
        className="br br-a150 quantity-box"
        data-jsx={3916862138}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        value="2"
      />
      <div
        className="license-info"
        data-jsx={3916862138}
      >
        <div
          data-jsx={3916862138}
        >
          Logan
           
          Howlett
        </div>
        <div
          data-jsx={3916862138}
          style={
            Object {
              "fontStyle": "italic",
            }
          }
        >
          4/15/2014
          <span
            className="id-label"
            data-jsx={3916862138}
          >
            ID:
          </span>
           
          000001
        </div>
      </div>
      <button
        data-jsx={3916862138}
        onClick={[Function]}
        type="button"
      >
        X
      </button>
    </div>
    <div
      className="p-a300 g"
      data-jsx={2595578495}
    >
      <div
        className="m-v500 g--9"
      >
        <div
          className="t--info"
        >
          7
           
          licenses
           × $
          14
           
          + 
          2.75
          % credit card fee
        </div>
        <div
          className="sh sh--b0"
        >
          <span
            className="sh-title"
          >
            Subtotal: $
            100.70
          </span>
        </div>
      </div>
      <a
        className="btn g--3 m-v500"
        data-jsx={2595578495}
        href="/dogs/payment"
        onClick={[Function]}
      >
        Pay and Finish
      </a>
    </div>
  </div>
</div>
`;

exports[`rendering renders loading for unhydrated cart 1`] = `
<div>
  <nav
    className="nv-s nv-s--sticky"
    data-jsx={792433154}
  >
    <div
      className="nv-s-l bar"
      data-jsx={792433154}
    >
      <a
        className="nv-s-l-b back-link"
        data-jsx={792433154}
        href="/dogs"
        onClick={[Function]}
      >
        Back to Lookup
      </a>
      <a
        className="cart-link"
        data-jsx={792433154}
        href="/dogs/checkout"
        onClick={[Function]}
      >
        8
      </a>
    </div>
  </nav>
  <div
    className="p-a300"
  >
    <div
      className="sh sh--b0 m-v300"
    >
      <h1
        className="sh-title"
      >
        Checkout
      </h1>
    </div>
  </div>
  <div>
    <div
      className="p-a300 g"
      data-jsx={2595578495}
    >
      <div
        className="m-v500 g--9"
      >
        <div
          className="t--info"
        >
          8
           
          licenses
           × $
          14
           
          + 
          2.75
          % credit card fee
        </div>
        <div
          className="sh sh--b0"
        >
          <span
            className="sh-title"
          >
            Subtotal: $
            115.09
          </span>
        </div>
      </div>
      <a
        className="btn g--3 m-v500"
        data-jsx={2595578495}
        href="/dogs/payment"
        onClick={[Function]}
      >
        Pay and Finish
      </a>
    </div>
  </div>
</div>
`;
